src/nsrt_learning.py

```diff
@@ -248,8 +244,102 @@ def _learn_pnad_side_predicates(
         ground_atom_dataset: List[GroundAtomTrajectory],
         train_tasks: List[Task], predicates: Set[Predicate],
         segments: List[Segment],
-        verbose: bool) -> List[PartialNSRTAndDatastore]:
-    raise NotImplementedError
+        segmented_trajs: List[List[Segment]]) -> List[PartialNSRTAndDatastore]:
+    # There are no goal states for this search; run until exhausted.
+    def _check_goal(s: Tuple[PartialNSRTAndDatastore, ...]) -> bool:
+        del s  # unused
+        return False
+
+    # Consider changing each effect to a side predicate.
+    def _get_successors(
+        s: Tuple[PartialNSRTAndDatastore, ...],
+    ) -> Iterator[Tuple[None, Tuple[PartialNSRTAndDatastore, ...], float]]:
+        for i in range(len(s)):
+            pnad = s[i]
+            _, option_vars = pnad.option_spec
+            for effect_set, add_or_delete in [(pnad.op.add_effects, "add"),
+                                              (pnad.op.delete_effects,
+                                               "delete")]:
+                for effect in effect_set:
+                    new_pnad = PartialNSRTAndDatastore(
+                        pnad.op.effect_to_side_predicate(
+                            effect, option_vars, add_or_delete),
+                        pnad.datastore, pnad.option_spec)
+                    sprime = list(s)
+                    sprime[i] = new_pnad
+                    yield (None, tuple(sprime), 1.0)
+
+    # Hard to avoid this circular import, since approaches need nsrt_learning.
+    from predicators.src.approaches.grammar_search_invention_approach import \
+        _PredictionErrorScoreFunction  # pylint:disable=import-outside-toplevel
+    score_func = _PredictionErrorScoreFunction(predicates, [], {}, train_tasks)
+
+    def _evaluate(s: Tuple[PartialNSRTAndDatastore, ...]) -> float:
+        strips_ops = [pnad.op for pnad in s]
+        option_specs = [pnad.option_spec for pnad in s]
+        score = score_func.evaluate_with_operators(frozenset(),
+                                                   ground_atom_dataset,
+                                                   segments, strips_ops,
+                                                   option_specs)
+        return score
+
+    # Run the search, starting from original PNADs.
+    path, _, _ = utils.run_hill_climbing(tuple(pnads), _check_goal,
+                                         _get_successors, _evaluate)
+    # Final PNADs are the last ones in the path.
+    pnads = list(path[-1])
+    # Recompute the datastores in the PNADs. We need to do this
+    # because now that we have side predicates, each transition may be
+    # assigned to *multiple* datastores.
+    _recompute_datastores_from_segments(segmented_trajs, pnads)
+    return pnads
+
+
+def _recompute_datastores_from_segments(
```

Comment:
why do we need to do this if we're reusing the machinery from grammar search? is the idea that the "pruned atom dataset" thing doesn't work here because it's not just about pruning atoms?

Reply:
this was your code! :P i think the comment above, which you wrote long ago, is clear enough:
```
    # Recompute the datastores in the PNADs. We need to do this
    # because now that we have side predicates, each transition may be
    # assigned to *multiple* datastores.
```

Reply:
I know, but I was wondering now that we're using the score function stuff, whether we could just reuse the existing code in the score function. but I think probably not, for the reason that I said above

Reply:
i think you are right